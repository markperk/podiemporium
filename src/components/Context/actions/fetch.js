const { REACT_APP_BASE_URL, REACT_APP_API_KEY } = process.env
const URL_DIRECTORY = '/api'

const standardHeaders = {
  "Content-Type": "application/json",
  "Accept": "application/json",
}

const getPostRequest = (urlPath, body) => {
  const init = {
    ...standardHeaders,
    'x-api-key': `${REACT_APP_API_KEY}`
  }
  const payload = {
    method: 'GET',
    headers: new Headers(init)
  }
  return promise(urlPath, payload, body)
}

const getRequest = urlPath => {
  const init = {
    ...standardHeaders,
    'x-api-key': `${REACT_APP_API_KEY}`
  }
  const payload = {
    method: 'GET',
    headers: new Headers(init)
  }
  return promise(urlPath, payload)
}

const getResponse = res => {
  if (res.ok && res.status === 204) {
    return Promise.resolve({
      status: res.status,
      message: 'No Content'
    })
  }
  if (res.ok) return res.json()
  return res.text()
    .then(error => JSON.parse(error))
    .then(error => {
      throw new Error(error.message)
    })
}

const promise = (urlPath, payload, body) => {
  return new Promise(resolve => {
    const url = `${REACT_APP_BASE_URL}${URL_DIRECTORY}${urlPath}`
    fetch(url, payload, body)
      .then(response => resolve(getResponse(response)))
      .catch(error => console.error('error', error))
  })
}

export const api = {
  get: (urlPath) => getRequest(urlPath),
  post: (urlPath, body) => getPostRequest(urlPath, body)
}