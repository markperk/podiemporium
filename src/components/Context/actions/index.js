import { api } from './fetch'

export {
  api,
}