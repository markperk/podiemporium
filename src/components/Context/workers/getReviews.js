import { api } from '../actions'

const assignPlayId = reviews => reviews.map(review => {
  review.play_id = Math.floor(Math.random() * 37) + 1
  return review
})

export const getReviews = async () => {
  return await api.get('/reviews').then(reviews => assignPlayId(reviews))
}