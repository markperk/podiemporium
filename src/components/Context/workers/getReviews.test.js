import { api } from '../actions/fetch'
import { getReviews } from './getReviews'

jest.mock('../actions/fetch')

const mockReviews = [{
  author: "Kaley Schiller",
  body: "The fool doth think he is wise, but the wise man knows himself to be a fool.",
  id: "9783221620868",
  play_id: 1,
  publish_date: "2016-09-05T23:25:47.642350Z",
  rating: 0.8,
}]

const { get } = api

describe('getReviews', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    global.Math.random = () => 1
    get.mockImplementation(() => {
      return new Promise(resolve => {
        resolve(mockReviews)
      })
    })

  })

  test('returns array of reviews and adds a random play_id to the review object', async () => {
    const result = await getReviews()
    expect(result).toMatchSnapshot()
  })
})