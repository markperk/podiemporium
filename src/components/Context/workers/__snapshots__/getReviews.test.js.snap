// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReviews returns array of reviews and adds a random play_id to the review object 1`] = `
Array [
  Object {
    "author": "Kaley Schiller",
    "body": "The fool doth think he is wise, but the wise man knows himself to be a fool.",
    "id": "9783221620868",
    "play_id": 38,
    "publish_date": "2016-09-05T23:25:47.642350Z",
    "rating": 0.8,
  },
]
`;
